File: HRMS/HRMS.Web/Modules/ServerTypes/Operations/LeaveTypeForm.ts
import { BooleanEditor, DecimalEditor, EnumEditor, initFormType, IntegerEditor, PrefixedContext, StringEditor, TextAreaEditor } from "@serenity-is/corelib";
import { LeaveCategory } from "./LeaveCategory";
import { RecordStatus } from "./RecordStatus";

export interface LeaveTypeForm {
    LeaveTypeName: StringEditor;
    LeaveCode: StringEditor;
    LeaveCategory: EnumEditor;
    AnnualAllocation: DecimalEditor;
    MonthlyAccrual: BooleanEditor;
    CarryForwardAllowed: BooleanEditor;
    MaxCarryForwardDays: DecimalEditor;
    EncashmentAllowed: BooleanEditor;
    ApplicableDepartments: TextAreaEditor;
    ApplicableDesignations: TextAreaEditor;
    ApplicableBranches: StringEditor;
    GenderSpecific: BooleanEditor;
    ProbationApplicable: BooleanEditor;
    MinimumServiceRequiredMonths: IntegerEditor;
    MaxLeavePerRequest: DecimalEditor;
    SandwichRuleApplicable: BooleanEditor;
    HalfDayAllowed: BooleanEditor;
    DocumentsRequired: BooleanEditor;
    Status: EnumEditor;
}

export class LeaveTypeForm extends PrefixedContext {
    static readonly formKey = 'Operations.LeaveType';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!LeaveTypeForm.init) {
            LeaveTypeForm.init = true;

            var w0 = StringEditor;
            var w1 = EnumEditor;
            var w2 = DecimalEditor;
            var w3 = BooleanEditor;
            var w4 = TextAreaEditor;
            var w5 = IntegerEditor;

            initFormType(LeaveTypeForm, [
                'LeaveTypeName', w0,
                'LeaveCode', w0,
                'LeaveCategory', w1,
                'AnnualAllocation', w2,
                'MonthlyAccrual', w3,
                'CarryForwardAllowed', w3,
                'MaxCarryForwardDays', w2,
                'EncashmentAllowed', w3,
                'ApplicableDepartments', w4,
                'ApplicableDesignations', w4,
                'ApplicableBranches', w0,
                'GenderSpecific', w3,
                'ProbationApplicable', w3,
                'MinimumServiceRequiredMonths', w5,
                'MaxLeavePerRequest', w2,
                'SandwichRuleApplicable', w3,
                'HalfDayAllowed', w3,
                'DocumentsRequired', w3,
                'Status', w1
            ]);
        }
    }
}

[LeaveCategory, RecordStatus]; // referenced types

File: HRMS/HRMS.Web/Modules/Operations/LeaveType/LeaveTypeForm.cs
using Serenity.ComponentModel;

namespace HRMS.Operations.Forms;

[FormScript("Operations.LeaveType")]
[BasedOnRow(typeof(LeaveTypeRow), CheckNames = true)]
public class LeaveTypeForm
{
    public string LeaveTypeName { get; set; }
    public string LeaveCode { get; set; }
    public LeaveCategory LeaveCategory { get; set; }
    public decimal AnnualAllocation { get; set; }
    public bool MonthlyAccrual { get; set; }
    public bool CarryForwardAllowed { get; set; }
    public decimal MaxCarryForwardDays { get; set; }
    public bool EncashmentAllowed { get; set; }
    [TextAreaEditor(Rows = 2)]
    public string ApplicableDepartments { get; set; }
    [TextAreaEditor(Rows = 2)]
    public string ApplicableDesignations { get; set; }
    public string ApplicableBranches { get; set; }
    public bool GenderSpecific { get; set; }
    public bool ProbationApplicable { get; set; }
    public int MinimumServiceRequiredMonths { get; set; }
    public decimal MaxLeavePerRequest { get; set; }
    public bool SandwichRuleApplicable { get; set; }
    public bool HalfDayAllowed { get; set; }
    public bool DocumentsRequired { get; set; }
    public RecordStatus Status { get; set; }
}
